### PR TITLE
added ibus.service file

### PR DESCRIPTION
**Pull request by shibumi**
Hello,
I have added an ibus.service file for systemd. Would be nice if you could merge this. 
The target-location for this file should be: `/lib/systemd/user/ibus.service`

### CODE REVIEW

**Comment by fujiwarat** [comment]
IBus has ibus/bus/org.freedesktop.IBus.service.in

**Comment by fujiwarat** [comment]
I don't think to run ibus with systemd since it does not get $XDG_CURRENT_DESKTOP.
Fedora uses https://tagoh.bitbucket.io/imsettings/ .

**Comment by shibumi** [comment]
@fujiwarat 
Is IBus.service.in a service file? it looks like a dbus interface...

So how do you start ibus-daemon at boot?

**Comment by fujiwarat** [comment]
> When I understand this right I can activate your dbus-activated service via dbus. But is there any service file that triggers this activation?

You're right. there is no trigger of ibus-daemon.

If you run a window manager, I think the trigger of the window manager also needs to run ibus-daemon.
I don't think using systemd is useful for the general desktops since it's out of the desktop processes for $XDG_CURRENT_DESKTOP.
Note: ibus-ui-gtk3 changes the UI by desktop.

**Comment by smcv** [comment]
> I don't think using systemd is useful for the general desktops since it's out of the desktop processes for $XDG_CURRENT_DESKTOP.

On recent versions of Debian, Fedora, Arch Linux etc., `dbus-daemon --session` is at least optionally started by `systemd --user` anyway. You can see this if you use `systemd-cgls` from a gnome-terminal: your terminal session is actually part of `gnome-terminal.service`. (In Debian you'd install the `dbus-user-session` package to opt-in to this, while in Fedora and Arch it's always enabled.)

 A well-integrated distribution should arrange for the currently active desktop session's `$XDG_CURRENT_DESKTOP` (and other relevant environment variables) to be copied into the environment used by `systemd --user` and `dbus-daemon --session`.

However, the proposed unit here was wrong because it duplicates the D-Bus service. The right thing to do is to link up the D-Bus service to the systemd service, which results in ibus-daemon being launched by `systemd --user` if that's how `dbus-daemon` was started, or directly by `dbus-daemon` otherwise (`dbus-daemon` knows that it should ignore the `SystemdService` and execute the `Exec` directly if it was not started by systemd). https://github.com/flatpak/xdg-desktop-portal-gtk/pull/52 demonstrates converting a service from simple D-Bus activation to (D-Bus + systemd) activation.

**Comment by fujiwarat** [comment]
Thank you for the kind explanation.
I guess imsettings also needs to use D-Bus + systemd activation.
It seems your suggestions are nice while I still don't understand the implementation with detail.

>  On recent versions of Debian, Fedora, Arch Linux etc., dbus-daemon --session is at least optionally started by systemd --user anyway.

Right. I think it causes a regression in the input methods.
Fedora uses imsettings to run ibus-daemon and it adds a workaround for this:
https://bitbucket.org/tagoh/imsettings/commits/7c759561511f4814bebfc0e1f5ba2459e560aaac
imsettings is launched by `systemd --user` but it uses the D-Bus communication between the server and clients.

> A well-integrated distribution should arrange for the currently active desktop session's $XDG_CURRENT_DESKTOP (and other relevant environment variables) to be copied into the environment used by systemd --user and dbus-daemon --session.

My understanding is, dbus-daemon copies some environment variables but not all and imsettings applied the above workaround.

**Comment by fujiwarat** [comment]
gnome-terminal also has a regression.
Currently changing the value of GTK_IM_MODULE likes `ibus` or `gtk-im-context-simple` does not effect gnom-terminal.
E.g. `env GTK_IM_MODULE=ibus gnome-terminal`

The workaround is to run /usr/libexec/gnome-terminal-server with GTK_IM_MODULE by manual without dbus before run gnome-terminal. E.g.
```
env GTK_IM_MODULE=gtk-im-context-simple /usr/libexec/gnome-terminal-server &
gnome-terminal
```

**Comment by smcv** [comment]
> My understanding is, dbus-daemon copies some environment variables but not all

`dbus-daemon` initially has whatever environment variables it inherited from its parent. In the "user sessions" setup, its parent is `systemd --user`. You can use `systemd-cgls` to see how things fit together.

Session startup programs (like `gnome-session`) or distribution integration scripts (`/etc/X11/Xsession.d` on Debian, `/etc/X11/xinit/xinitrc.d` on Fedora and others, `/etc/profile.d` for login shells) can copy some or all environment variables into the environment that `dbus-daemon --session` and `systemd --user` will give to their child processes, either by using the `dbus-update-activation-environment` command-line tool (more suitable for shell scripts) or by calling the `UpdateActivationEnvironment()` D-Bus method on the `dbus-daemon` (more suitable for C/Python/etc. code).

Services with only a D-Bus `.service` file in `/usr/share/dbus-1/services` are child processes of the `dbus-daemon`, and are started with whatever environment the `dbus-daemon` has been given.

Services with a D-Bus `.service` file linked via `SystemdService` to a systemd `.service` file in `/usr/lib/systemd/user` are child processes of `systemd --user`, and are started with whatever environment `systemd --user` has been given, unless that environment is overridden on a per-service basis. `systemd --user` is a better service manager than `dbus-daemon` is, because that's what it's designed to be, whereas `dbus-daemon` only starts services as a secondary function.

This might be a clearer explanation: https://bugzilla.gnome.org/show_bug.cgi?id=755735#c3

**Comment by fujiwarat** [comment]
I mean the environment variables has the different times:
https://bugzilla.gnome.org/show_bug.cgi?id=738205#c29

**Comment by smcv** [comment]
> I mean the environment variables has the different timings: https://bugzilla.gnome.org/show_bug.cgi?id=738205#c29

Yes, but you already have this problem on systems that start `dbus-daemon --session` from `systemd --user`, whether you have your own systemd service or not.

That is being worked on, but it is going to have to be solved in something more central - gnome-session, distro integration scripts or systemd - and not in individual services like ibus.

**Comment by shibumi** [comment]
@smcv @fujiwarat 
I have updated my PR.. It's now similar to the PR: https://github.com/flatpak/xdg-desktop-portal-gtk/pull/52
But i have one last question. Can I use this service as user-enabled service? so with "systemctl --user start ibus.service"?? Or how does this work now.. I have to agree with @fujiwarat .. this dbus-systemd-connection is somehow complicated.

**Comment by smcv** [comment]
> Can I use this service as user-enabled service? so with "systemctl --user start ibus.service"?

Yes. You can't `systemctl enable` it unless you add an `[Install]` section, though.

**Comment by fujiwarat** [comment]
Sorry, I forgot how DBus treats the environment variables and noted the wrong comment:

>>  dbus-daemon copies some environment variables but not all

I'm getting remembering the implementation and actually it's a timing issue when dbus-daemon is launched above.

> That is being worked on, but it is going to have to be solved in something more central - gnome-session, distro integration scripts or systemd - and not in individual services like ibus.

My understanding is, the resolution is to send some environment variables with DBus methods between the DBus clients and servers but IBus still needs other environment variables.
And the workaround was also applied to the imsettings.

> Yes, but you already have this problem on systems that start dbus-daemon --session from systemd --user, whether you have your own systemd service or not.

Are you suggesting the patch of this pull request?
ibus/bus/org.freedesktop.IBus.service.in was added for gnome-shell but seems gnome-shell still runs ibus-daemon directly in gnome-shell/js/misc/ibusManager.js so I don't know if the org.freedesktop.IBus.service is actually used.

**Comment by shibumi** [comment]
@fujiwarat @smcv 
I have updated the service file. It will propagate the ibus socket now, this way we can activate the ibus daemon via a user activated service file and applications can dock on the ibus.socket.

Thats my homesetup:

```
● ibus.service - ibus daemon
   Loaded: loaded (/etc/systemd/user/ibus.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2017-03-07 17:32:44 CET; 22min ago
     Docs: man:ibus-daemon(1)
 Main PID: 19798 (ibus-daemon)
   CGroup: /user.slice/user-1000.slice/user@1000.service/ibus.service
           ├─19798 /usr/bin/ibus-daemon --replace --xim -a unix:tmpdir=$IBUS_SOCK
           ├─19801 /usr/lib/ibus/ibus-dconf
           ├─19802 /usr/lib/ibus/ibus-ui-gtk3
           ├─19805 /usr/lib/ibus/ibus-x11 --kill-daemon
           └─19869 /usr/lib/ibus/ibus-engine-simple

Mar 07 17:32:44 motoko systemd[435]: Starting ibus daemon...
Mar 07 17:32:44 motoko systemd[435]: Started ibus daemon.
```

```
# /etc/systemd/user/ibus.service
[Unit]
Description=ibus daemon
Documentation=man:ibus-daemon(1)

[Service]
Type=dbus
BusName=org.freedesktop.IBus
Environment=IBUS_SOCK=%t/ibus.socket
ExecStart=/usr/bin/ibus-daemon --replace --xim -a unix:tmpdir=$IBUS_SOCK
```

```
# /home/chris/.zshrc.pre
export EDITOR=vim
export SSH_AUTH_SOCK="$XDG_RUNTIME_DIR/ssh-agent.socket"
export IBUS_SOCK="$XDG_RUNTIME_DIR/ibus.socket"
```

**Comment by fujiwarat** [comment]
CC'ing @rtcm
@rtcm I think org.freedesktop.IBus.service.in file is not used in gnome-shell yet?
Basically I don't think to integrate the patch for the special desktop environment.

@shibumi Why do you integrate your patch in ibus git?
The patch does not have any install system paths so I think you can have your files personally.

**Comment by rtcm** [comment]
@fujiwarat indeed, I intended to change gnome-shell to activate ibus-daemon through DBus after the service file landed in IBus but moved on to other things and forgot about it.

I think having a systemd service file is actually a better and more modern way to achieve what I wanted in the first place. We could remove the DBus service file and replace it with a systemd one.

One problem I see with this is that currently ibus-daemon only accepts some options via command line switches like --xim and --panel. I think we'd need a way for ibus clients to request those features be activated dynamically at run time so that the service itself can be started generically without command line switches and then configured by the specific desktop environment.

File: ibus.service.in [hunk 1]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=ibus daemon
+Documentation=man:ibus-daemon(1)
+
+[Service]
+Type=dbus
+BusName=org.freedesktop.IBus
+Environment=IBUS_SOCK=%t/ibus.socket
+ExecStart=@bindir@/ibus-daemon --replace --xim --panel disable -a unix:tmpdir=$IBUS_SOCK
```
Comment (fujiwarat) [inline ibus.service.in]: Please replace bus/org.freedesktop.IBus.service.in  with bus/bus.service.in.
Please replace '--xim' with @xim_cli_arg@
Please modify bus/Makefile.am with the install path.

I wonder if the ibus.service file can get an environment variable likes ibus-daemon --xim $IBUS_ARGS or can have a user directory likes $XDG_CONFIG_HOME/systemd/ibus.service

Comment (shibumi) [inline ibus.service.in]: Hello, why should org.freedesktop.IBus.service.in renamed to bus.service.in? the service file in `bus/` is a `dbus-activated`-service. Therefore it should have a FQDN service name. Just `bus.service.in` is not specific enough.I would further go with `ibus.service` as well in the `/etc/systemd/user/`-dir.

For ibus.service you can use `/etc/systemd/user/ibus.service` or `$HOME/.config/systemd/user/ibus.service`.

I will modify the part with `--xim` this makes absolutly sense.

Comment (fujiwarat) [inline ibus.service.in]: OK, thank you for the explanation. And the both  looks good.

File: ibus.service.in [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=ibus daemon
+Documentation=man:ibus-daemon(1)
+
+[Service]
+Type=dbus
+BusName=org.freedesktop.IBus
+Environment=IBUS_SOCK=%t/ibus.socket
```
Comment (fujiwarat) [inline ibus.service.in]: Since it's a directory name, IBUS_SOCKET=%t/ibus/socket is better?

Comment (fujiwarat) [inline ibus.service.in]: Thinking this again, %t may be better than %t/ibus.socket since I cannot find mkdir

**Comment by shibumi** [comment]
Mhh I have still issues with this commit. When I login the service starts, but it seems like I can't control ibus via my specified keys. When I restart the service manually it's working fine.. very strange.

**Comment by rtlanceroad** [comment]
I think this totally should be done by distro side rather than upstream... Why spread virus all over the place?

**Comment by shibumi** [comment]
@rtlanceroad what has this to do with malware? And no, this has to be done upstream, service files should be part of every upstream project to keep them consistent and secure.

**Comment by rtlanceroad** [comment]
The virus, I mean systemd...
Service files are _only_ for your init system, and input methods should keep independent from any init system, don't you think so?  So if you want it,  cook it yourself, and don't pollute upstream...

**Comment by shibumi** [comment]
@rtlanceroad We shouldn't discuss this in this issue.. You don't seem to understand what systemd is and have an absolutly wrong view about it.

**Comment by rtlanceroad** [comment]
Yeah, we shouldn't discuss it. Systemd gave me bad experience, caused one of my pcs unbootable in the past, that was why i switched my system to non-systemd distro. Anyway i insist that input methods should be independent from any init and your thing should be done in your side.  Your work just causes packagers like me to do extra work(remove those) every new release if it's merged.

**Comment by fujiwarat** [comment]
@shibumi 
> When I login the service starts, but it seems like I can't control ibus via my specified keys.

What is your problem?

**Comment by ryutaroh-matsumoto** [comment]
I don't know if `ibus.service` is going to be included in a next release, but if it is, the following entries might deserve inclusion:

```
ExecStop=/usr/bin/ibus exit
ExecReload=/usr/bin/ibus reset
```
cc: @fujiwarat

**Comment by fujiwarat** [comment]
@shibumi Your problem can be fixed with @ryutaroh-matsumoto 's suggestion?

> ExecReload=/usr/bin/ibus reset

I guess you mean `ibus restart`.

**Comment by ryutaroh-matsumoto** [comment]
Sorry I was wrong. "reset" should have been "restart".

**Comment by matsumoto-returned** [comment]
When I start ibus-daemon outside of any GUI i.e. from console login, I must run `ibus engine mozc-jp` or something like that to activate some engine. I regard this as a feature of IBus as it seems to be expected to be started under some GUI or a desktop environment.

**Comment by fujiwarat** [comment]
When I looked at /usr/lib/systemd/user/gnome-shell.service, it has `After` and `PartOf` keywords.
I think ibus-daemon needs to run after the desktop system to get the workspace size instead of the monitor size and I guess the 'After' keyword could be defined by desktop system and I'm not sure if the service file is useful.
You might run ibus-daemon with systemd-run ?

**Comment by ohk2kt3t4** [comment]
```
Environment=IBUS_SOCK=%t/ibus.socket
ExecStart=/usr/bin/ibus-daemon --replace --xim -a unix:tmpdir=$IBUS_SOCK
```
When running ibus-daemon as a user service, what is the significance of specifically setting the socket file here instead of just using the default one created by ibus-daemon in /tmp?

**Comment by fujiwarat** [comment]
I think the systemd file will be fixed by https://github.com/ibus/ibus/pull/2381

**Comment by shibumi** [comment]
Totally, forgot that this still exists. Sorry everybody. I wrote my bachelor thesis between 2018/2019 :D might make sense why I did not respond. Good to see that @fujiwarat fixed this

**Pull request closed without merging**